sklearn/metrics/pairwise.py

```diff
@@ -551,7 +567,253 @@ def cosine_distances(X, Y=None):
+            lo = np.nanmin(col_array)
+            hi = np.nanmax(col_array)
+
+            if y_is_not_none:
+                col_array = Y_num[:, col]
+                lo = np.nanmin([lo, np.nanmin(col_array)])
+                hi = np.nanmax([hi, np.nanmax(col_array)])
+
+            if np.isnan(lo):
+                lo = 0.0
+            if np.isnan(hi):
+                hi = 0.0
+
+            min_of_numeric[col] = lo
+            max_of_numeric[col] = hi
+
+        # Scales the numeric values between 0 and 1.
+        if scale:
+            scaler = MinMaxScaler().fit([min_of_numeric, max_of_numeric])
+            X_num = scaler.transform(X_num)
+            if y_is_not_none:
+                Y_num = scaler.transform(Y_num)
+            else:
+                Y_num = X_num
+        # If not, checks if the data is scaled.
+        elif np.min(min_of_numeric) < 0 or np.max(max_of_numeric) > 1:
+            warnings.warn("Input data is not scaled between 0 and 1.",
+                          UserWarning)
+
+    y_n_rows = Y.shape[0]
+    D = np.zeros((n_rows, y_n_rows), dtype=np.float)
+
+    for i in range(n_rows):
+        j_start = i
+
+        # for non square results
+        if n_rows != y_n_rows or y_is_not_none:
+            j_start = 0
+
+        sum_s = 0
+        n_cols_not_missing = np.zeros(n_rows - j_start)
+        # sum and missing values treatment for each type of data, if available
+        if n_col_cat_obj_present:
+            row_cat_none = ~(np.equal(X_cat_obj[i, :], None) |
+                             np.equal(Y_cat_obj[j_start:n_rows, :], None))
+
+            n_cols_not_missing = np.sum(row_cat_none.astype(np.int), axis=1)
+            sum_s = (~np.equal(X_cat_obj[i, :], Y_cat_obj[j_start:n_rows, :])
+                     & row_cat_none).astype(np.int).sum(axis=1)
+
+        if n_col_cat_num_present:
+            row_cat_nan = ~(np.isnan(X_cat_num[i, :]) |
+                            np.isnan(Y_cat_num[j_start:n_rows, :]))
+
+            n_cols_not_missing = \
+                n_cols_not_missing + \
+                np.sum(row_cat_nan.astype(np.int), axis=1)
+
+            sum_s = sum_s + (~np.equal(X_cat_num[i, :],
+                                       Y_cat_num[j_start:n_rows, :])
+                             & row_cat_nan).astype(np.int).sum(axis=1)
+
+        if n_col_num_present:
+            n_cols_not_missing = n_cols_not_missing + \
+                                 np.sum((~(np.isnan(X_num[i, :]) |
+                                           np.isnan(Y_num[j_start:n_rows, :])))
+                                        .astype(np.int), axis=1)
+
+            sum_s = sum_s + np.nansum(
+                np.absolute(X_num[i, :] - Y_num[j_start:n_rows, :]), axis=1)
+
+        result = np.divide(sum_s, n_cols_not_missing,
+                           out=np.full(n_cols_not_missing.shape, np.nan),
+                           where=n_cols_not_missing != 0)
+
+        D[i, j_start:] = result
```

Comment:
If you run a non-squared distance matrix with only numeric features defined then this line will throw a ValueError.
e.g.
```python
X = np.array([[1.0, 0.0, 0.0], [0.181818, 0.0, 1], [0.0, 0.0, 0.160377]], dtype=object)
Y = np.array([[0.090909, 0.0, 0.500109]], dtype=object)

D = gower_distances(X, Y)
```